File: clients/algoliasearch-client-php/lib/Model/QuerySuggestions/SourceIndiceWithReplicas.php
<?php

namespace Algolia\AlgoliaSearch\Model\QuerySuggestions;

/**
 * SourceIndiceWithReplicas Class Doc Comment
 *
 * @category Class
 * @description Source indice with replicas used to generate a Query Suggestions index.
 *
 * @package Algolia\AlgoliaSearch
 */
class SourceIndiceWithReplicas extends \Algolia\AlgoliaSearch\Model\AbstractModel implements ModelInterface, \ArrayAccess, \JsonSerializable
{
    /**
      * Array of property to type mappings. Used for (de)serialization
      *
      * @var string[]
      */
    protected static $modelTypes = [
        'replicas' => 'bool',
        'indexName' => 'string',
        'analyticsTags' => 'string[]',
        'facets' => 'object[]',
        'minHits' => 'int',
        'minLetters' => 'int',
        'generate' => 'string[][]',
        'external' => '\Algolia\AlgoliaSearch\Model\QuerySuggestions\SourceIndexExternal[]',
    ];

    /**
      * Array of property to format mappings. Used for (de)serialization
      *
      * @var string[]
      */
    protected static $modelFormats = [
        'replicas' => null,
        'indexName' => null,
        'analyticsTags' => null,
        'facets' => null,
        'minHits' => null,
        'minLetters' => null,
        'generate' => null,
        'external' => null,
    ];

    /**
     * Array of property to type mappings. Used for (de)serialization
     *
     * @return array
     */
    public static function modelTypes()
    {
        return self::$modelTypes;
    }

    /**
     * Array of property to format mappings. Used for (de)serialization
     *
     * @return array
     */
    public static function modelFormats()
    {
        return self::$modelFormats;
    }

    /**
     * Array of attributes to setter functions (for deserialization of responses)
     *
     * @var string[]
     */
    protected static $setters = [
        'replicas' => 'setReplicas',
        'indexName' => 'setIndexName',
        'analyticsTags' => 'setAnalyticsTags',
        'facets' => 'setFacets',
        'minHits' => 'setMinHits',
        'minLetters' => 'setMinLetters',
        'generate' => 'setGenerate',
        'external' => 'setExternal',
    ];

    /**
     * Array of attributes to getter functions (for serialization of requests)
     *
     * @var string[]
     */
    protected static $getters = [
        'replicas' => 'getReplicas',
        'indexName' => 'getIndexName',
        'analyticsTags' => 'getAnalyticsTags',
        'facets' => 'getFacets',
        'minHits' => 'getMinHits',
        'minLetters' => 'getMinLetters',
        'generate' => 'getGenerate',
        'external' => 'getExternal',
    ];

    /**
     * Array of attributes to setter functions (for deserialization of responses)
     *
     * @return array
     */
    public static function setters()
    {
        return self::$setters;
    }

    /**
     * Array of attributes to getter functions (for serialization of requests)
     *
     * @return array
     */
    public static function getters()
    {
        return self::$getters;
    }

    /**
     * Associative array for storing property values
     *
     * @var mixed[]
     */
    protected $container = [];

    /**
     * Constructor
     *
     * @param mixed[] $data Associated array of property values
     */
    public function __construct(array $data = null)
    {
        if (isset($data['replicas'])) {
            $this->container['replicas'] = $data['replicas'];
        }
        if (isset($data['indexName'])) {
            $this->container['indexName'] = $data['indexName'];
        }
        if (isset($data['analyticsTags'])) {
            $this->container['analyticsTags'] = $data['analyticsTags'];
        }
        if (isset($data['facets'])) {
            $this->container['facets'] = $data['facets'];
        }
        if (isset($data['minHits'])) {
            $this->container['minHits'] = $data['minHits'];
        }
        if (isset($data['minLetters'])) {
            $this->container['minLetters'] = $data['minLetters'];
        }
        if (isset($data['generate'])) {
            $this->container['generate'] = $data['generate'];
        }
        if (isset($data['external'])) {
            $this->container['external'] = $data['external'];
        }
    }

    /**
     * Show all the invalid properties with reasons.
     *
     * @return array invalid properties with reasons
     */
    public function listInvalidProperties()
    {
        $invalidProperties = [];

        if (!isset($this->container['replicas']) || $this->container['replicas'] === null) {
            $invalidProperties[] = "'replicas' can't be null";
        }
        if (!isset($this->container['indexName']) || $this->container['indexName'] === null) {
            $invalidProperties[] = "'indexName' can't be null";
        }
        if (!isset($this->container['analyticsTags']) || $this->container['analyticsTags'] === null) {
            $invalidProperties[] = "'analyticsTags' can't be null";
        }
        if (!isset($this->container['facets']) || $this->container['facets'] === null) {
            $invalidProperties[] = "'facets' can't be null";
        }
        if (!isset($this->container['minHits']) || $this->container['minHits'] === null) {
            $invalidProperties[] = "'minHits' can't be null";
        }
        if (!isset($this->container['minLetters']) || $this->container['minLetters'] === null) {
            $invalidProperties[] = "'minLetters' can't be null";
        }
        if (!isset($this->container['generate']) || $this->container['generate'] === null) {
            $invalidProperties[] = "'generate' can't be null";
        }
        if (!isset($this->container['external']) || $this->container['external'] === null) {
            $invalidProperties[] = "'external' can't be null";
        }

        return $invalidProperties;
    }

    /**
     * Validate all the properties in the model
     * return true if all passed
     *
     * @return bool True if all properties are valid
     */
    public function valid()
    {
        return count($this->listInvalidProperties()) === 0;
    }

    /**
     * Gets replicas
     *
     * @return bool
     */
    public function getReplicas()
    {
        return $this->container['replicas'] ?? null;
    }

    /**
     * Sets replicas
     *
     * @param bool $replicas true if the Query Suggestions index is a replicas
     *
     * @return self
     */
    public function setReplicas($replicas)
    {
        $this->container['replicas'] = $replicas;

        return $this;
    }

    /**
     * Gets indexName
     *
     * @return string
     */
    public function getIndexName()
    {
        return $this->container['indexName'] ?? null;
    }

    /**
     * Sets indexName
     *
     * @param string $indexName source index name
     *
     * @return self
     */
    public function setIndexName($indexName)
    {
        $this->container['indexName'] = $indexName;

        return $this;
    }

    /**
     * Gets analyticsTags
     *
     * @return string[]
     */
    public function getAnalyticsTags()
    {
        return $this->container['analyticsTags'] ?? null;
    }

    /**
     * Sets analyticsTags
     *
     * @param string[] $analyticsTags list of analytics tags to filter the popular searches per tag
     *
     * @return self
     */
    public function setAnalyticsTags($analyticsTags)
    {
        $this->container['analyticsTags'] = $analyticsTags;

        return $this;
    }

    /**
     * Gets facets
     *
     * @return object[]
     */
    public function getFacets()
    {
        return $this->container['facets'] ?? null;
    }

    /**
     * Sets facets
     *
     * @param object[] $facets list of facets to define as categories for the query suggestions
     *
     * @return self
     */
    public function setFacets($facets)
    {
        $this->container['facets'] = $facets;

        return $this;
    }

    /**
     * Gets minHits
     *
     * @return int
     */
    public function getMinHits()
    {
        return $this->container['minHits'] ?? null;
    }

    /**
     * Sets minHits
     *
     * @param int $minHits Minimum number of hits (e.g., matching records in the source index) to generate a suggestions.
     *
     * @return self
     */
    public function setMinHits($minHits)
    {
        $this->container['minHits'] = $minHits;

        return $this;
    }

    /**
     * Gets minLetters
     *
     * @return int
     */
    public function getMinLetters()
    {
        return $this->container['minLetters'] ?? null;
    }

    /**
     * Sets minLetters
     *
     * @param int $minLetters minimum number of required letters for a suggestion to remain
     *
     * @return self
     */
    public function setMinLetters($minLetters)
    {
        $this->container['minLetters'] = $minLetters;

        return $this;
    }

    /**
     * Gets generate
     *
     * @return string[][]
     */
    public function getGenerate()
    {
        return $this->container['generate'] ?? null;
    }

    /**
     * Sets generate
     *
     * @param string[][] $generate List of facet attributes used to generate Query Suggestions. The resulting suggestions are every combination of the facets in the nested list (e.g., (facetA and facetB) and facetC).
     *
     * @return self
     */
    public function setGenerate($generate)
    {
        $this->container['generate'] = $generate;

        return $this;
    }

    /**
     * Gets external
     *
     * @return \Algolia\AlgoliaSearch\Model\QuerySuggestions\SourceIndexExternal[]
     */
    public function getExternal()
    {
        return $this->container['external'] ?? null;
    }

    /**
     * Sets external
     *
     * @param \Algolia\AlgoliaSearch\Model\QuerySuggestions\SourceIndexExternal[] $external list of external indices to use to generate custom Query Suggestions
     *
     * @return self
     */
    public function setExternal($external)
    {
        $this->container['external'] = $external;

        return $this;
    }
    /**
     * Returns true if offset exists. False otherwise.
     *
     * @param int $offset Offset
     *
     * @return bool
     */
    public function offsetExists($offset)
    {
        return isset($this->container[$offset]);
    }

    /**
     * Gets offset.
     *
     * @param int $offset Offset
     *
     * @return mixed|null
     */
    public function offsetGet($offset)
    {
        return $this->container[$offset] ?? null;
    }

    /**
     * Sets value based on offset.
     *
     * @param int|null $offset Offset
     * @param mixed    $value  Value to be set
     *
     * @return void
     */
    public function offsetSet($offset, $value)
    {
        if (is_null($offset)) {
            $this->container[] = $value;
        } else {
            $this->container[$offset] = $value;
        }
    }

    /**
     * Unsets offset.
     *
     * @param int $offset Offset
     *
     * @return void
     */
    public function offsetUnset($offset)
    {
        unset($this->container[$offset]);
    }
}

File: clients/algoliasearch-client-java-2/algoliasearch-core/src/com/algolia/model/querySuggestions/SourceIndiceWithReplicas.java
package com.algolia.model.querySuggestions;

import com.google.gson.annotations.SerializedName;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

/** Source indice with replicas used to generate a Query Suggestions index. */
public class SourceIndiceWithReplicas {

  @SerializedName("replicas")
  private Boolean replicas;

  @SerializedName("indexName")
  private String indexName;

  @SerializedName("analyticsTags")
  private List<String> analyticsTags = new ArrayList<>();

  @SerializedName("facets")
  private List<Object> facets = new ArrayList<>();

  @SerializedName("minHits")
  private Integer minHits;

  @SerializedName("minLetters")
  private Integer minLetters;

  @SerializedName("generate")
  private List<List<String>> generate = new ArrayList<>();

  @SerializedName("external")
  private List<SourceIndexExternal> external = new ArrayList<>();

  public SourceIndiceWithReplicas setReplicas(Boolean replicas) {
    this.replicas = replicas;
    return this;
  }

  /**
   * true if the Query Suggestions index is a replicas.
   *
   * @return replicas
   */
  @javax.annotation.Nonnull
  public Boolean getReplicas() {
    return replicas;
  }

  public SourceIndiceWithReplicas setIndexName(String indexName) {
    this.indexName = indexName;
    return this;
  }

  /**
   * Source index name.
   *
   * @return indexName
   */
  @javax.annotation.Nonnull
  public String getIndexName() {
    return indexName;
  }

  public SourceIndiceWithReplicas setAnalyticsTags(List<String> analyticsTags) {
    this.analyticsTags = analyticsTags;
    return this;
  }

  public SourceIndiceWithReplicas addAnalyticsTagsItem(
    String analyticsTagsItem
  ) {
    this.analyticsTags.add(analyticsTagsItem);
    return this;
  }

  /**
   * List of analytics tags to filter the popular searches per tag.
   *
   * @return analyticsTags
   */
  @javax.annotation.Nonnull
  public List<String> getAnalyticsTags() {
    return analyticsTags;
  }

  public SourceIndiceWithReplicas setFacets(List<Object> facets) {
    this.facets = facets;
    return this;
  }

  public SourceIndiceWithReplicas addFacetsItem(Object facetsItem) {
    this.facets.add(facetsItem);
    return this;
  }

  /**
   * List of facets to define as categories for the query suggestions.
   *
   * @return facets
   */
  @javax.annotation.Nonnull
  public List<Object> getFacets() {
    return facets;
  }

  public SourceIndiceWithReplicas setMinHits(Integer minHits) {
    this.minHits = minHits;
    return this;
  }

  /**
   * Minimum number of hits (e.g., matching records in the source index) to generate a suggestions.
   *
   * @return minHits
   */
  @javax.annotation.Nonnull
  public Integer getMinHits() {
    return minHits;
  }

  public SourceIndiceWithReplicas setMinLetters(Integer minLetters) {
    this.minLetters = minLetters;
    return this;
  }

  /**
   * Minimum number of required letters for a suggestion to remain.
   *
   * @return minLetters
   */
  @javax.annotation.Nonnull
  public Integer getMinLetters() {
    return minLetters;
  }

  public SourceIndiceWithReplicas setGenerate(List<List<String>> generate) {
    this.generate = generate;
    return this;
  }

  public SourceIndiceWithReplicas addGenerateItem(List<String> generateItem) {
    this.generate.add(generateItem);
    return this;
  }

  /**
   * List of facet attributes used to generate Query Suggestions. The resulting suggestions are
   * every combination of the facets in the nested list (e.g., (facetA and facetB) and facetC).
   *
   * @return generate
   */
  @javax.annotation.Nonnull
  public List<List<String>> getGenerate() {
    return generate;
  }

  public SourceIndiceWithReplicas setExternal(
    List<SourceIndexExternal> external
  ) {
    this.external = external;
    return this;
  }

  public SourceIndiceWithReplicas addExternalItem(
    SourceIndexExternal externalItem
  ) {
    this.external.add(externalItem);
    return this;
  }

  /**
   * List of external indices to use to generate custom Query Suggestions.
   *
   * @return external
   */
  @javax.annotation.Nonnull
  public List<SourceIndexExternal> getExternal() {
    return external;
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) {
      return true;
    }
    if (o == null || getClass() != o.getClass()) {
      return false;
    }
    SourceIndiceWithReplicas sourceIndiceWithReplicas = (SourceIndiceWithReplicas) o;
    return (
      Objects.equals(this.replicas, sourceIndiceWithReplicas.replicas) &&
      Objects.equals(this.indexName, sourceIndiceWithReplicas.indexName) &&
      Objects.equals(
        this.analyticsTags,
        sourceIndiceWithReplicas.analyticsTags
      ) &&
      Objects.equals(this.facets, sourceIndiceWithReplicas.facets) &&
      Objects.equals(this.minHits, sourceIndiceWithReplicas.minHits) &&
      Objects.equals(this.minLetters, sourceIndiceWithReplicas.minLetters) &&
      Objects.equals(this.generate, sourceIndiceWithReplicas.generate) &&
      Objects.equals(this.external, sourceIndiceWithReplicas.external)
    );
  }

  @Override
  public int hashCode() {
    return Objects.hash(
      replicas,
      indexName,
      analyticsTags,
      facets,
      minHits,
      minLetters,
      generate,
      external
    );
  }

  @Override
  public String toString() {
    StringBuilder sb = new StringBuilder();
    sb.append("class SourceIndiceWithReplicas {\n");
    sb.append("    replicas: ").append(toIndentedString(replicas)).append("\n");
    sb
      .append("    indexName: ")
      .append(toIndentedString(indexName))
      .append("\n");
    sb
      .append("    analyticsTags: ")
      .append(toIndentedString(analyticsTags))
      .append("\n");
    sb.append("    facets: ").append(toIndentedString(facets)).append("\n");
    sb.append("    minHits: ").append(toIndentedString(minHits)).append("\n");
    sb
      .append("    minLetters: ")
      .append(toIndentedString(minLetters))
      .append("\n");
    sb.append("    generate: ").append(toIndentedString(generate)).append("\n");
    sb.append("    external: ").append(toIndentedString(external)).append("\n");
    sb.append("}");
    return sb.toString();
  }

  /**
   * Convert the given object to string with each line indented by 4 spaces (except the first line).
   */
  private String toIndentedString(Object o) {
    if (o == null) {
      return "null";
    }
    return o.toString().replace("\n", "\n    ");
  }
}
